### PR TITLE
sleek: init at 1.3.1

### PR DESCRIPTION
**Pull request by auroraanna**
###### Description of changes

Added package sleek.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by auroraanna** [comment]
It currently only works if you run do `steam-run sleek`. If you don't do that it doesn't even tell you what the problem is.
```console
$ ./result/bin/.sleek-wrapped
zsh: no such file or directory: ./result/bin/.sleek-wrapped
```

**Comment by auroraanna** [comment]
This is only x86_64-linux for now. I plan to add other platforms later.

File: pkgs/applications/office/sleek/snap/gui/sleek.desktop.patch [hunk 1]
```diff
@@ -0,0 +1,16 @@
+diff --git a/snap/gui/sleek.desktop b/snap/gui/sleek.desktop
+index ce43c15..decea7d 100755
+--- a/snap/gui/sleek.desktop
++++ b/snap/gui/sleek.desktop
+@@ -1,9 +1,9 @@
+ [Desktop Entry]
+ Name=sleek
+-Exec=sleek %U
++Exec=sleek
```
Comment (SuperSandro2000) [inline pkgs/applications/office/sleek/snap/gui/sleek.desktop.patch:9]: Why not forward args?

File: pkgs/applications/office/sleek/default.nix [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  buildPhase = ''
+    yarn --offline build:linux
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -vp  $out/opt $out/bin $out/share/{icons,applications}
+    cp -r deps/${pname}/dist/linux-unpacked/* $out/opt/
+    ln -s $out/opt/${pname} $out/bin/
```
Comment (SuperSandro2000) [inline pkgs/applications/office/sleek/default.nix:52]: Please do not substitute pname wherever possible

File: pkgs/applications/office/sleek/default.nix [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    #cp -r sleek $out/bin/
+    #cp -r chrome* icudtl.dat LICENSE* locales resources* snapshot_blob.bin v* $out/share/
+    #cp -r lib* swiftshader $out/lib/
```
Comment (SuperSandro2000) [inline pkgs/applications/office/sleek/default.nix:55]:
```suggestion
```

File: pkgs/applications/office/sleek/default.nix [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  postPatch = ''
+    substituteInPlace src/scss/style.scss --replace '../../node_modules' $node_modules
+    substituteInPlace package.json --replace 'yarn' 'yarn --offline'
+
+    export HOME=$(mktemp -d)
+
+    mkdir -p $HOME/.cache/{electron,electron-builder/fpm/fpm-1.9.3-2.3.1-linux-x86_64}
+    ln -s ${electron} $HOME/.cache/electron/electron-v18.3.7-linux-x64.zip
+    ln -s ${fpm}/bin/fpm $HOME/.cache/electron-builder/fpm/fpm-1.9.3-2.3.1-linux-x86_64/fpm
```
Comment (SuperSandro2000) [inline pkgs/applications/office/sleek/default.nix:40]: Can we easily get those version numbers or use wildcards to make updates easier?

File: pkgs/applications/office/sleek/default.nix [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  electron = fetchurl {
+    url = "https://github.com/electron/electron/releases/download/v18.3.7/electron-v18.3.7-linux-x64.zip";
+    hash = "sha256-4JjmZB6HWjBZhRiwtFDMKsG8/nfvK/1jkvIxJU+DUos=";
+  };
+
+  nativeBuildInputs = [ wrapGAppsHook ];
+
+  patches = [
+    ./package.json.patch
+    ./snap/gui/sleek.desktop.patch
```
Comment (SuperSandro2000) [inline pkgs/applications/office/sleek/default.nix:29]:
```suggestion
    ./sleek.desktop.patch
```

File: pkgs/applications/office/sleek/default.nix [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  src = fetchFromGitHub {
+    owner = "ransome1";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-j5T4StCcHtNHEn0VycPJlX/w+F0jZYBPEgHI/n5nGiI=";
+  };
+
+  electron = fetchurl {
+    url = "https://github.com/electron/electron/releases/download/v18.3.7/electron-v18.3.7-linux-x64.zip";
```
Comment (SuperSandro2000) [inline pkgs/applications/office/sleek/default.nix:21]: Please use the electron from nixpkgs

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/office/sleek/snap/gui/sleek.desktop.patch (hunk 1 above)
Comment (auroraanna) [reply pkgs/applications/office/sleek/snap/gui/sleek.desktop.patch:9]: You're right. I didn't understand what it does yesterday and made a stupid decision.

File: pkgs/applications/office/sleek/default.nix (hunk 2 above)
Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:52]: I thought it may reduce spelling errors…

Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:52]: It's removed now.

File: pkgs/applications/office/sleek/default.nix (hunk 3 above)
Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:55]: oops, left the comment

File: pkgs/applications/office/sleek/default.nix (hunk 5 above)
Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix]: @auroraanna didn't properly read the first line of the patch and thought the folder it's in matters.

File: pkgs/applications/office/sleek/default.nix (hunk 4 above)
Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:40]: This is not needed anymore with nixpkgs electron.

File: pkgs/applications/office/sleek/default.nix (hunk 6 above)
Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:21]: Yep!

**Comment by auroraanna** [comment]
Due to using nixpkgs electron every issue but that electron_18 is EOL has been fixed.

**Comment by auroraanna** [comment]
https://github.com/ransome1/sleek/pull/481

**Comment by auroraanna** [comment]
I don't understand why ofborg fails.

**Comment by auroraanna** [comment]
https://github.com/ransome1/sleek/issues/482 related to the `Exec=sleek %U` in the .desktop file.

**Comment by auroraanna** [comment]
Another issue, bitwarden and jitsi-meet-electron put `--add-flags "\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations}}"` in their makeWrapper command but I tried those programs on sway and xeyes tells me they are still running on Xorg. How to run on wayland?

**Comment by auroraanna** [comment]
Should this be moved to nodePackages?

**Comment by SuperSandro2000** [comment]
> Should this be moved to nodePackages?

No, updating nodePackages is a pain.

File: pkgs/applications/office/sleek/default.nix [hunk 7]
```diff
@@ -0,0 +1,71 @@
+{ mkYarnPackage
+, fetchFromGitHub
+, electron_18
+, wrapGAppsHook
+, lib
+}:
+
+mkYarnPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/applications/office/sleek/default.nix:8]: I think mkYarnPackage is missing its package.json.

**Comment by auroraanna** [comment]
> > Should this be moved to nodePackages?
> 
> No, updating nodePackages is a pain.

But, should it be moved there at some point when updating nodePackages is not a pain anymore?

**Comment by SuperSandro2000** [comment]
> But, should it be moved there at some point when updating nodePackages is not a pain anymore?

I don't think there are plans for that right now. nodePackages would need to be redone with a different tool or the tool would need major improvements.

File: pkgs/applications/office/sleek/default.nix (hunk 7 above)
Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:8]: But the package.json is in the src, it's even being patched successfully. What do you mean it's missing?

**Comment by auroraanna** [comment]
> [ransome1/sleek#481](https://github.com/ransome1/sleek/pull/481)

It seems to work perfectly fine with electron 23 though. If none responds to that issue, should this package just use electron 23?

File: pkgs/applications/office/sleek/default.nix (hunk 7 above)
Comment (SuperSandro2000) [reply pkgs/applications/office/sleek/default.nix:8]: I think that is triggering IFD which is forbidden in nixpkgs

**Comment by auroraanna** [comment]
I added the update script.

**Comment by auroraanna** [comment]
The checkPhase fails and I don't understand the error.

File: pkgs/applications/office/sleek/default.nix (hunk 7 above)
Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:9]: I'm missing context. Why is it triggering IFD?

Comment (SuperSandro2000) [reply pkgs/applications/office/sleek/default.nix:9]: mkYarnPackage falls internally back to IFD. We should probably disable that for nixpkgs. Please see other packages how to provide the package.json.

Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:9]: Did I do it correctly in the latest commit?

Comment (SuperSandro2000) [reply pkgs/applications/office/sleek/default.nix:9]: ofborg eval still fails with the following error which means IFD is used

```
nix-env failed:
error:
       … while evaluating call site

       at «none»:0: (source not available)

       … while calling anonymous lambda

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/.gc-of-borg-outpaths.nix:47:12:

           46|   tweak = lib.mapAttrs
           47|     (name: val:
             |            ^
           48|       if name == "recurseForDerivations" then true

       … while evaluating call site

       at «none»:0: (source not available)

       … while calling 'g'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/attrsets.nix:595:19:

          594|           g =
          595|             name: value:
             |                   ^
          596|             if isAttrs value && cond value

       … while evaluating call site

       at «none»:0: (source not available)

       … while calling anonymous lambda

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/top-level/release-lib.nix:148:38:

          147|      set of meta.platforms values. */
          148|   packagePlatforms = mapAttrs (name: value:
             |                                      ^
          149|       if isDerivation value then

       … while evaluating the attribute 'meta.hydraPlatforms'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/stdenv/generic/make-derivation.nix:558:13:

          557|      inherit passthru overrideAttrs;
          558|      inherit meta;
             |             ^
          559|    } //

       … while evaluating call site

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/stdenv/generic/make-derivation.nix:505:10:

          504|
          505|   meta = checkMeta.commonMeta { inherit validity attrs pos references; };
             |          ^
          506|   validity = checkMeta.assertValidity { inherit meta attrs; };

       … while calling 'commonMeta'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/stdenv/generic/check-meta.nix:401:16:

          400|   #   validity = checkMeta.assertValidity { inherit meta attrs; };
          401|   commonMeta = { validity, attrs, pos ? null, references ? [ ] }:
             |                ^
          402|     let

       … while evaluating the attribute 'meta'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix:391:7:

          390|
          391|       meta = {
             |       ^
          392|         inherit (nodejs.meta) platforms;

       … while evaluating call site

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix:393:12:

          392|         inherit (nodejs.meta) platforms;
          393|       } // lib.optionalAttrs (package ? description) { inherit (package) description; }
             |            ^
          394|         // lib.optionalAttrs (package ? homepage) { inherit (package) homepage; }

       … while calling 'optionalAttrs'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/attrsets.nix:675:5:

          674|     # The attribute set to return if `cond` is `true`.
          675|     as:
             |     ^
          676|     if cond then as else {};

       … while evaluating call site

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix:269:17:

          268|     let
          269|       package = lib.importJSON packageJSON;
             |                 ^
          270|       pname = package.name;

       … while calling 'importJSON'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/trivial.nix:312:16:

          311|   */
          312|   importJSON = path:
             |                ^
          313|     builtins.fromJSON (builtins.readFile path);

       … while realising the context of path '/nix/store/mbhbjyk01mlpakmwlzh673hkyi6lhxar-source/package.json'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/trivial.nix:313:24:

          312|   importJSON = path:
          313|     builtins.fromJSON (builtins.readFile path);
             |                        ^
          314|

       error: path '/nix/store/nj2j3nn03j2p5zwv7yb8d29rihhkhdjq-source.drv' is not valid

```

Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:9]: oh, IFD is used before it's converting the JSON to nix and then IFDing that. I don't know how else to specify the package.json though. A lot of other node programs do that.

Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:9]: the entire package.json would have to be converted to nix and then inlined into the default.nix

Comment (auroraanna) [reply pkgs/applications/office/sleek/default.nix:9]: /me is looking at more other yarn packages

**Comment by DamienCassou** [comment]
Any update on that?

**Comment by auroraanna** [comment]
superseded by #271676

**Pull request closed without merging**